File: advancedscreenswitchersample/src/main/java/me/aartikov/advancedscreenswitchersample/ui/MainActivity.java
package me.aartikov.advancedscreenswitchersample.ui;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.aartikov.advancedscreenswitchersample.R;
import me.aartikov.advancedscreenswitchersample.SampleApplication;
import me.aartikov.advancedscreenswitchersample.SampleScreenSwitcherAnimationProvider;
import me.aartikov.advancedscreenswitchersample.SampleTransitionAnimationProvider;
import me.aartikov.advancedscreenswitchersample.screens.MainScreen;
import me.aartikov.advancedscreenswitchersample.screens.TabScreen;
import me.aartikov.alligator.NavigationContext;
import me.aartikov.alligator.NavigationContextBinder;
import me.aartikov.alligator.Navigator;
import me.aartikov.alligator.Screen;
import me.aartikov.alligator.annotations.RegisterScreen;
import me.aartikov.alligator.listeners.ScreenSwitchingListener;
import me.aartikov.alligator.screenswitchers.FragmentScreenSwitcher;

/**
 * Date: 21.01.2016
 * Time: 23:30
 *
 * @author Artur Artikov
 */
@RegisterScreen(MainScreen.class)
public class MainActivity extends AppCompatActivity implements BottomNavigationView.OnNavigationItemSelectedListener, ScreenSwitchingListener {
	@BindView(R.id.bottom_bar)
	BottomNavigationView mBottomBar;

	private Navigator mNavigator = SampleApplication.getNavigator();
	private NavigationContextBinder mNavigationContextBinder = SampleApplication.getNavigationContextBinder();
	private FragmentScreenSwitcher mScreenSwitcher;

	@SuppressLint("UseSparseArrays")
	private Map<Integer, Screen> mTabScreenMap = new LinkedHashMap<>();

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		ButterKnife.bind(this);

		initTabScreenMap();
		mBottomBar.setOnNavigationItemSelectedListener(this);
		mScreenSwitcher = new FragmentScreenSwitcher(SampleApplication.getNavigationFactory(), getSupportFragmentManager(),
				R.id.main_container, new SampleScreenSwitcherAnimationProvider(getTabScreens()));

		if (savedInstanceState == null) {
			mNavigator.switchTo(getTabScreen(R.id.tab_android));
		}
	}

	private void initTabScreenMap() {
		mTabScreenMap.put(R.id.tab_android, new TabScreen(getString(R.string.tab_android)));
		mTabScreenMap.put(R.id.tab_bug, new TabScreen(getString(R.string.tab_bug)));
		mTabScreenMap.put(R.id.tab_dog, new TabScreen(getString(R.string.tab_dog)));
	}

	@Override
	protected void onResumeFragments() {
		super.onResumeFragments();
		bindNavigationContext();
	}

	@Override
	protected void onPause() {
		mNavigationContextBinder.unbind();
		super.onPause();
	}

	private void bindNavigationContext() {
		NavigationContext.Builder builder = new NavigationContext.Builder(this)
				.screenSwitcher(mScreenSwitcher)
				.screenSwitchingListener(this)
				.transitionAnimationProvider(new SampleTransitionAnimationProvider());

		Fragment fragment = mScreenSwitcher.getCurrentFragment();
		if (fragment != null && fragment instanceof ContainerIdProvider) {
			builder.containerId(((ContainerIdProvider) fragment).getContainerId())
					.fragmentManager(fragment.getChildFragmentManager());       // Use child fragment manager for nested navigation
		}

		mNavigationContextBinder.bind(builder.build());
	}

	@Override
	public void onBackPressed() {
		mNavigator.goBack();
	}

	@Override
	public boolean onNavigationItemSelected(@NonNull MenuItem item) {
		Screen screen = getTabScreen(item.getItemId());
		mNavigator.switchTo(screen);
		return false;
	}

	@Override
	public void onScreenSwitched(@Nullable Screen screenFrom, @NonNull Screen screenTo) {
		int tabId = getTabId(screenTo);
		mBottomBar.getMenu().findItem(tabId).setChecked(true);
		bindNavigationContext();    // rebind NavigationContext because we need to set another container id and another child fragment manager.
	}

	private Screen getTabScreen(int tabId) {
		return mTabScreenMap.get(tabId);
	}

	private int getTabId(Screen tabScreen) {
		for (Map.Entry<Integer, Screen> entry : mTabScreenMap.entrySet()) {
			if (tabScreen.equals(entry.getValue())) {
				return entry.getKey();
			}
		}
		return -1;
	}

	private List<Screen> getTabScreens() {
		return new ArrayList<>(mTabScreenMap.values());
	}
}

File: advancedscreenswitchersample/src/main/java/me/aartikov/advancedscreenswitchersample/ui/MainActivity.kt
package me.aartikov.advancedscreenswitchersample.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.bottomnavigation.BottomNavigationView
import me.aartikov.advancedscreenswitchersample.R
import me.aartikov.advancedscreenswitchersample.SampleApplication
import me.aartikov.advancedscreenswitchersample.SampleScreenSwitcherAnimationProvider
import me.aartikov.advancedscreenswitchersample.SampleTransitionAnimationProvider
import me.aartikov.advancedscreenswitchersample.screens.MainScreen
import me.aartikov.advancedscreenswitchersample.screens.TabScreen
import me.aartikov.alligator.NavigationContext
import me.aartikov.alligator.NavigationContextBinder
import me.aartikov.alligator.Navigator
import me.aartikov.alligator.Screen
import me.aartikov.alligator.annotations.RegisterScreen
import me.aartikov.alligator.listeners.ScreenSwitchingListener
import me.aartikov.alligator.screenswitchers.FragmentScreenSwitcher

@RegisterScreen(MainScreen::class)
class MainActivity : AppCompatActivity(),
    BottomNavigationView.OnNavigationItemSelectedListener,
    ScreenSwitchingListener {

    private lateinit var mBottomBar: BottomNavigationView

    private val mNavigator: Navigator = SampleApplication.navigator
    private val mNavigationContextBinder: NavigationContextBinder = SampleApplication.navigationContextBinder
    private lateinit var mScreenSwitcher: FragmentScreenSwitcher

    @SuppressLint("UseSparseArrays")
    private val mTabScreenMap: MutableMap<Int, Screen> = LinkedHashMap()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mBottomBar = findViewById(R.id.bottom_bar)

        initTabScreenMap()
        mBottomBar.setOnItemSelectedListener(this)
        mScreenSwitcher = FragmentScreenSwitcher(
            SampleApplication.navigationFactory,
            supportFragmentManager,
            R.id.main_container, SampleScreenSwitcherAnimationProvider(tabScreens)
        )

        if (savedInstanceState == null) {
            mNavigator.switchTo(getTabScreen(R.id.tab_android))
        }
    }

    private fun initTabScreenMap() {
        mTabScreenMap[R.id.tab_android] = TabScreen(getString(R.string.tab_android))
        mTabScreenMap[R.id.tab_bug] = TabScreen(getString(R.string.tab_bug))
        mTabScreenMap[R.id.tab_dog] = TabScreen(getString(R.string.tab_dog))
    }

    override fun onResumeFragments() {
        super.onResumeFragments()
        bindNavigationContext()
    }

    override fun onPause() {
        mNavigationContextBinder.unbind(this)
        super.onPause()
    }

    private fun bindNavigationContext() {
        val builder = NavigationContext.Builder(this, SampleApplication.navigationFactory)
            .screenSwitcher(mScreenSwitcher)
            .screenSwitchingListener(this)
            .transitionAnimationProvider(SampleTransitionAnimationProvider())

        val fragment = mScreenSwitcher.currentFragment
        if (fragment is ContainerIdProvider) {
            builder.fragmentNavigation(
                fragment.childFragmentManager,
                (fragment as ContainerIdProvider).containerId
            ) // Use child fragment manager for nested navigation
        }

        mNavigationContextBinder.bind(builder.build())
    }

    @SuppressLint("MissingSuperCall")
    override fun onBackPressed() {
        mNavigator.goBack()
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        val screen = getTabScreen(item.itemId)
        mNavigator.switchTo(screen)
        return false
    }

    override fun onScreenSwitched(screenFrom: Screen?, screenTo: Screen) {
        val tabId = getTabId(screenTo)
        mBottomBar.menu.findItem(tabId).setChecked(true)
        bindNavigationContext() // rebind NavigationContext because we need to set another container id and another child fragment manager.
    }

    private fun getTabScreen(tabId: Int): Screen {
        return mTabScreenMap.getValue(tabId)
    }

    private fun getTabId(tabScreen: Screen): Int {
        for ((key, value) in mTabScreenMap) {
            if (tabScreen == value) {
                return key
            }
        }
        return -1
    }

    private val tabScreens: List<Screen>
        get() = ArrayList(mTabScreenMap.values)
}
